libafl_qemu/src/qemu/config.rs

```diff
@@ -350,40 +345,28 @@ pub struct QemuConfig {
     program: Program,
 } // Adding something here? Please leave Program as the last field
 
-impl From<QemuConfig> for Result<Qemu, QemuInitError> {
-    /// This method is necessary to make the API resemble a typical builder pattern, i.e.
-    /// `Qemu::builder().foo(bar).build()`, while still leveraging `TypedBuilder` for this
-    /// non-standard use case where `Qemu` doesn't store the configuration.
-    /// Internally, `TypedBuilder` is used to generate a builder for `QemuConfig`.
-    /// This `QemuConfig.into()` method is used by the derived `QemuConfigBuilder.build()`
-    /// to go from `QemuConfigBuilder` to `QemuConfig`, and finally to `Qemu` in one fn.
-    ///
-    /// # Errors
-    /// returns `QemuInitError` if the Qemu initialization fails, including cases where Qemu has
-    /// already been initialized.
-    fn from(config: QemuConfig) -> Self {
-        let args = config
+impl From<&QemuConfig> for Vec<String> {
+    /// Generate the QEMU-compatible initialization cli string from the QEMU config.
+    fn from(config: &QemuConfig) -> Self {
+        config
             .to_string()
             .split(' ')
             .map(ToString::to_string)
-            .collect::<Vec<String>>();
-        let qemu = Qemu::init(&args)?;
-        QEMU_CONFIG
-            .set(config)
-            .map_err(|_| unreachable!("BUG: QEMU_CONFIG was already set but Qemu was not init!"))?;
-        Ok(qemu)
+            .collect::<Vec<String>>()
     }
 }
 
 #[cfg(test)]
 mod test {
     use super::*;
+    use crate::Qemu;
 
     #[test]
     #[cfg(feature = "usermode")]
     fn usermode() {
         let program = "/bin/pwd";
-        let qemu = Qemu::builder().program("/bin/pwd").build().unwrap();
```

Comment:
Doens't the old API look better here? Can't we just keep Qemu::builder() and `.build()` then calls `Qemu::init_whatever`?

Reply:
That's true, but the thing is that we need to separate the configuration step from the init step, and the builder generated by `typed-builder` is not made for this, it's a pain to move around

Reply:
IMHO it's more important to have a good API than to have nice code "on the inside".
This is already what you're using (right?) https://github.com/idanarye/rust-typed-builder/pull/80

Reply:
from the user's point of view, i think separating configuration from actual initialization is not so bad as well.
it makes more clear what the configuration part is and where qemu is actually getting initialized.

hiding something like qemu initialization in a `From` implementation doesn't sound good imho

Reply:
No it's really bad. We have builders everywhere in the codebase and they all work the same. There's no reason this shoudl work differently. You can still pass around an un-built builder

Reply:
but it works like any other builder, right? it's just that the builder doesn't build directly qemu, but qemu's config instead.
i think `QemuConfig::builder` makes it quite clear what we are building is not QEMU, right?

Reply:
Yes IMHO it should directly build qemu, not build a builder lol. Why build a builder?

Reply:
What I think this should be is a Qemu::builder not a QemuConfig::builder, I don't get what the indirection is for

Reply:
right, i see your point. i think i have an idea to make it more idiomatic.
it's a bit more complex than usual here because it's possible to build `Qemu` from a `&[String]` and a `QemuConfig`, which are not completely compatible.
the latest modification to `Qemu::init` should make it possible to simplify this and have one builder.

Reply:
i think there is something we cannot solve with the way `Emulator` works now.
the main issue is that now, we must make sure  `Emulator` is the one that initializes `Qemu` if we want the pre-initialization hooks to run.
so we must give to `EmulatorBuilder` something to build Qemu (something equivalent to the current `QemuConfig`) and let `Emulator` use it to initialize `Qemu`.
if we make `QemuConfigBuilder` build `Qemu` directly (like it was done before), we cannot make `Emulator` run the hooks correctly.
so i think we must necessarily have an intermediate representation of Qemu's config that can be used to initialize qemu (`QemuConfig`, so).
what i can do is create a `QemuBuilder` that mimics the old behavior or `QemuConfigBuilder` and calls `Qemu::init` internally tho, even if we can't really use it with `Emulator` if ppl really want to initialize `Qemu` themselves.

Reply:
I think in general we should *either* have Qemu _or_ Emulator, it's super hard to understand why we have both things right now. Then you also only need one builder..

Reply:
Otherwise maybe just change Emulator to take an initialized Qemu if it's a wrapper around Qemu?

Reply:
Hey @rmalmain, about:
> we must necessarily have an intermediate representation of Qemu's config that can be used to initialize qemu

As @domenukk  was saying, how hard would it be to pass a `QemuConfigBuilder` to `Emulator`? `Emulator` can then call `QemuConfigBuilder.build() -> Qemu` after the Hooks and stuff setup.

@domenukk  About
> Otherwise maybe just change Emulator to take an initialized Qemu if it's a wrapper around Qemu?

It was like this before, the problem is that
> we must make sure Emulator is the one that initializes Qemu if we want the pre-initialization hooks to run.

Reply:
Why not migrate all uses of Qemu to Emulator directly? I (as a user) don't get the split between those two

Reply:
@domenukk the idea of `Qemu` is to have a 0-sized object giving a slightly higher-level interface to QEMU (basically making it more convenient to call to QEMU functions and providing a more rusty api), that can only be obtained when QEMU is correctly initialized.
i use this trick a lot to give a `Qemu` as parameter for 0 cost.
also (i may be wrong), i think it's nice to let the possibility to use QEMU without all the things in `Emulator` and still provide a usable API.

in practice, i think you're right and most people will not want to use `Qemu` alone at all. the separation mostly exists to make it simpler to separate things and make it more clear what is specific to QEMU and what is higher-level stuff we introduce. A possible thing we can try is to move `Qemu` and `QemuHooks` to `libafl_qemu_sys`? at least it would remove the confusion between `Emulator` and `Qemu` for users.

@Marcondiro  i thought about this, but i think it will be quite heavy to do in practice because `rust-typed-builder` adds up many generics internally that would be difficult to annotate explicitly.

Reply:
Why not make a builder for emulator instead, and remove qemu from the user-facing APIs then?

Reply:
Also, you're not married to typed_builder if it dropping it makes the API cleaner

Reply:
`QemuConfig` is quite big, not using something like `typed_builder` would be a lot of work and make it less maintainable. idk of any other typed builder like this one.
`EmulatorBuilder` already exists btw, it's used to set things like the modules, snapshot engine, etc...
`QemuConfig` is just a programmatic way to set the cli of QEMU. if we continue to use the derive macro for getting qemu cli from QemuConfig, we need all the fields of the builder to be an option of the QEMU cli, hence the current split.